src/pghstore/_speedups.c

```diff
@@ -0,0 +1,373 @@
+  while (p != NULL) { /* loop through all the c's */
+    q = p; /* scan backwards through preceding escapes */
+    while (q > s && *(q-1) == '\\')
+      q--;
+    if (isEven(p-q)) /* even number of esc's => c is good */
+        return p;
+    p = strchr(p+1, c); /* else odd escapes => c is escaped, keep going */
+  }
+  return NULL;
+}
+
+/* Written to match pghstore._native.unescape but slightly smarter. */
+char *
+unescape(char *start_of_string, char *end_of_string)
+{
+  ssize_t copy_index = 0;
+  ssize_t index;
+  ssize_t string_length = (end_of_string - start_of_string) + 1;
+  /* NOTE(sigmavirus24): In the event that there are \s in the string, we're
+   * over-allocating here but that is okay since we're going to free shortly
+   * after use.
+   */
+  char *unescaped_string = calloc(string_length, sizeof(char));
+
+  for (index = 0; index < (string_length - 1); index++) {
+      if (start_of_string[index] == '\\' && start_of_string[index + 1] != '\\') {
+          continue;
+      }
+      unescaped_string[copy_index++] = start_of_string[index];
+  }
+
+  return unescaped_string;
+}
+
+static PyObject *
+_speedups_loads(PyObject *self, PyObject *args, PyObject *keywds)
+{
+  static char *keyword_argument_names[] = {"string", "encoding", "return_type", NULL};
+  const char *errors = "strict";
+  char *encoding = "utf-8";
+  char *s;
+  Py_ssize_t s_length = 0;
+  int i, null_value = 0;
+  char *unescaped_key, *unescaped_value;
+  char *key_start, *key_end, *value_start, *value_end;
+  PyTypeObject *return_type = &PyDict_Type;
+  PyObject *return_value, *key, *value;
+
+  if (!PyArg_ParseTupleAndKeywords(args, keywds, "s#|sO", keyword_argument_names, &s, &s_length, &encoding, &return_type)) {
+    return NULL;
+  }
+  /* NOTE(sigmavirus24): Because we use `s#` to parse the input string, we
+   * also need to pass s_length to get the length of the string because `s`
+   * may have embedded null characters.
+   * All of our tests presently pass but it's plausible that we could find
+   * data with null characters inside and have to update this to match.
+   */
+
+  return_value = PyObject_CallObject((PyObject *) return_type, NULL);
+  
+  // Each iteration will find one key/value pair
+  while ((key_start = strchr(s, '"')) != NULL) {
+    // move to the next character after the found "
+    key_start++;
+
+    // Find end of key
+    key_end = strchr_unescaped(key_start, '"');
+
+    // Find begining of value or NULL
+    for (i=1; 1; i++) {
+      switch (*(key_end+i)) {
+      case 'N':
+      case 'n':
+        // found NULL
+        null_value = 1;
+        break;
+      case '"':
+        // found begining of value
+        value_start = key_end+i+1;
+        break;
+      case '\0':
+        // found end of string
+        return return_value;
+      default:
+        // neither NULL nor begining of value, keep looking
+        continue;
+      }
+      break;
+    }
+     
+    unescaped_key = unescape(key_start, key_end);
```

Comment:
Well since we're yack shaving, do we need the unecoded string anywhere but as a decoded python unicode string? If not why not make unencode return the python string then we're not duplicating a bunch of code *and* you get to have the calloc and free in the same function.